### ISSUE TITLE
utils

### ISSUE THREAD

**Issue by khalil030306-dotcom** [isaacveg/Strange_Little_Things#3]
perplexity.py
`import re
import pandas as pd
import matplotlib.pyplot as plt

def parse_log_file(file_path):
    """
    Parses a log file to extract iteration and PPL values.
    """
    data = []
    with open(file_path, 'r') as f:
        for line in f:
            match = re.search(r'iteration (\d+) \|.*lm loss PPL: ([\d\.E\+]+)', line)
            if match:
                iteration = int(match.group(1))
                ppl = float(match.group(2))
                data.append({'iteration': iteration, 'ppl': ppl})
    return pd.DataFrame(data)

def plot_ppl(log_files):
    """
    Plots PPL vs. iteration from multiple log files.
    """
    plt.figure(figsize=(10, 6))

    for log_file in log_files:
        df = parse_log_file(log_file)
        if not df.empty:
            plt.plot(df['iteration'], df['ppl'], marker='o', linestyle='-', label=log_file)

    plt.title('PPL vs. Iteration')
    plt.xlabel('Iteration')
    plt.ylabel('PPL (Perplexity)')
    plt.xlim(10000,16000)
    plt.ylim(15, 25)
    plt.grid(True)
    plt.legend()
    plt.savefig('ppl_vs_iteration3.png')
    plt.close()

# List of your log files
# log_files = [ '/data01/xuxinchen/crosstrain/code/logs/mindspeed/train_llama_150m_diloco_h=100.log',
#               '/data01/xuxinchen/crosstrain/code/logs/mindspeed/train_llama_150m_cdiloco.log',
#               '/data01/xuxinchen/crosstrain/code/logs/mindspeed/train_llama_150m_sdiloco_s=5_h=100.log']

log_files = [ '/data01/xuxinchen/crosstrain/code/logs/mindspeed/train_llama_150m_sdiloco_s=5_h=100.log',
              '/data01/xuxinchen/crosstrain/code/logs/mindspeed/train_llama_150m_sdiloco_s=4.log']

# Generate the plot
plot_ppl(log_files)

print("Plot saved as ppl_vs_iteration.png")`

**Comment by khalil030306-dotcom** [comment]
parse_log.py
`import re
import csv
import os

def parse_log_to_csv(log_file_paths, output_csv_paths):
    """
    从日志文件中解析训练数据并保存为 CSV 文件。

    参数:
    log_file_path (str): 输入的日志文件路径。
    output_csv_path (str): 输出的 CSV 文件路径。
    """
    # 正则表达式，用于匹配并捕获所需的数据
    # - (\d+): 捕获迭代次数 (整数)
    # - ([\d.E\+\-]+): 捕获损失值和PPL (支持浮点数和科学计数法)
    training_pattern = re.compile(
        r"elapsed time per iteration \(ms\): ([\d.]+)"
    )
    log_pattern = re.compile(
        r"validation loss at iteration (\d+) \| "
        r"lm loss value: ([\d.E\+\-]+) \| "
        r"lm loss PPL: ([\d.E\+\-]+) \|"
    )
    cumulative_time_ms = 0.0
    # 存储提取到的数据
    extracted_data = []

    # print(f"正在读取日志文件: {log_file_path}...")
    for log_file_path, output_csv_path in zip(log_file_paths, output_csv_paths):
        try:
            with open(log_file_path, 'r', encoding='utf-8') as f:
                for line in f:
                    training_match = training_pattern.search(line)
                    if training_match:
                        elapsed_ms = float(training_match.group(1))
                        cumulative_time_ms += elapsed_ms
                        continue # 处理完这行就继续下一行
                    match = log_pattern.search(line)
                    if match:
                        # 提取匹配到的三个组
                        iteration = int(match.group(1))
                        loss = float(match.group(2))
                        perplexity = float(match.group(3))
                        
                        # 将提取的数据添加到列表中
                        extracted_data.append([iteration, loss, perplexity, cumulative_time_ms])
        except FileNotFoundError:
            print(f"错误: 未找到日志文件 '{log_file_path}'。请确保文件存在且路径正确。")
            return
        except Exception as e:
            print(f"读取文件时发生错误: {e}")
            return

        if not extracted_data:
            print("未在日志文件中找到任何匹配的数据行。")
            return

        print(f"成功提取 {len(extracted_data)} 条数据。")

        # 将数据写入 CSV 文件
        try:
            with open(output_csv_path, 'w', newline='', encoding='utf-8') as csv_file:
                writer = csv.writer(csv_file)
                
                # 写入表头
                writer.writerow(['iteration', 'loss', 'perplexity', 'total_elapsed_time_ms'])
                
                # 写入所有提取的数据
                writer.writerows(extracted_data)
            
            print(f"数据已成功写入到: {output_csv_path}")

        except Exception as e:
            print(f"写入 CSV 文件时发生错误: {e}")
        extracted_data = []


# --- 使用说明 ---

if __name__ == "__main__":
    
    # 2. 定义输入和输出文件名
    input_log_file = ["train_llama_150m_cdiloco_s=1.log", "train_llama_150m_cdiloco_s=4_N=5.log", "train_llama_150m_diloco_h=100.log", "train_llama_150m_sdiloco_s=5_h=100.log", "train_llama_150m_sdiloco_s=1.log"]  # 你的日志文件名
    output_csv_file = ["train_llama_150m_cdiloco_s=1.csv", "train_llama_150m_cdiloco_s=4_N=5.csv", "train_llama_150m_diloco_h=100.csv", "train_llama_150m_sdiloco_s=5_h=100.csv", "train_llama_150m_sdiloco_s=1.csv"] # 你希望生成的CSV文件名

    # 3. 运行解析函数
    parse_log_to_csv(input_log_file, output_csv_file)`
